reference/best_practices/reusable_bundles.rst

```diff
@@ -1,4 +1,95 @@
+
+
+Compatibility
+-------------
+
+Your bundle won't be reusable for some customers if you forgot some cases. But that's not a problem if they are aware of the compatibility of your bundle.
+
+Akeneo PIM storage stack comes with MySQL and hybrid MySQL/MongoDB (:doc:`/reference/technical_information/choose_database`).
+Use the ProductQueryBuilder which is an overlayer making work both storages with the same API.
+The following cookbook has been done to help you: ":doc:`/cookbook/catalog/product/query`".
+
+Two existing distributions CE/EE. The main project is `Akeneo PIM Community Edition`_ (CE) but
+we also have an `Enterprise Edition`_ (EE) which comes with a bunch of extra features dedicated to advanced needs.
+
+Many attribute types are managed (14 in CE and one more in EE). Be sure you don't forget a use case for an attribute type.
+
+Try your bundle on another catalogs. There is a multiplicity of catalog possibilities.
+Don't forget use cases with many locales, many channels and various catalog volumes.
+You can use our `data generator bundle`_ to generate a catalog.
+
+.. _Akeneo PIM Community Edition: https://github.com/akeneo/pim-community-standard
+.. _Enterprise Edition: https://www.akeneo.com/enterprise-edition/
+.. _data generator bundle: https://github.com/akeneo-labs/DataGeneratorBundle
+
+
+Maintainability
+---------------
+
+First, keep in mind the `Single Responsibility` from the five SOLID principles of Oriented Object Programming.
+It seems simple but we often forget to explode our code in many thinner classes.
+The code is easily readable and testable.
+
+You can use `PMD`_ to help you to make your code simple, flexible and easily testable.
+
+
+Now your code should be testable, so test your code! :)
+In Akeneo PIM, we use `PhpSpec`_ as unit testing tool and `Behat`_ for to test our application.
+
+Use a Continuous Integration to deploy your application, test your code and improve its quality.
+On our open-source projects, we use free CI tools `Travis`_ and `Scrutinizer`_
+You can take some examples on our `Akeneo-Labs`_ projects
+
+.. _PMD: https://phpmd.org/
+.. _PhpSpec: http://phpspec.readthedocs.org/
+.. _Behat: http://docs.behat.org/
+.. _Travis: https://travis-ci.org/
+.. _Scrutinizer: https://scrutinizer-ci.com/
+.. _Akeneo-Labs: https://github.com/akeneo-labs
+
+
+Extensibility
+-------------
+
+Reusability
+
+Avoid to override Akeneo PIM classes/services
+-> avoid to override class parameter
+-> avoid to add properties to entities
+
+--> create your own service (with your own class if you have to add specific code) for your use case.
+It won't impact the whole Akeneo PIM application
+
+--> Use composition instead of extension
+
+
+
+Create and rely on interfaces
+Plug on events (datagrid, etc.)
+Use Registries
+
+
+Attribute types could be added
+
+
+
+Scalability
+-----------
+
+know your limits and document them
+
+Check the memory usage
+Use php-meminfo
```

Comment:
php-meminfo is of course a very good tool yet for very specific purposes. Sometimes, µprofiler/Blackfire.io make more sense, so you should add them to this list.